Errors and Panic in Rueidis during Redis Cluster Shard Failover

### Description

We observed a set of issues when using rueidis v1.0.69 with a Redis Cluster (v7.4).

During a failover in one shard of the cluster, the application experienced an increase in Redis request latency, followed by multiple client-side errors. No application code changes were deployed during this time.

The following issues were observed:

- Intermittent transaction errors:
```
EXEC was aborted by redis or connection closed
```

- Redis protocol parse errors:
```
rueidis: parse error: redis message type simple string is not a array
rueidis: parse error: redis message type array is not a string
```

- Connection-level errors:
```
EOF
dial tcp <ip>:<port>: operation was canceled
```

Once the parse errors started, the client did not recover and required restarting application pods.

### Panic observed during cluster refresh
In the same time window, we observed a panic originating from the cluster topology refresh path.
```
panic: runtime error: index out of range [2] with length 0

goroutine 258842186 [running]:
github.com/redis/rueidis.parseSlots(...)
    /go/pkg/mod/github.com/redis/rueidis@v1.0.69/cluster.go:378
github.com/redis/rueidis.clusterslots.parse(...)
    /go/pkg/mod/github.com/redis/rueidis@v1.0.69/cluster.go:378
github.com/redis/rueidis.(*clusterClient)._refresh(...)
    /go/pkg/mod/github.com/redis/rueidis@v1.0.69/cluster.go:167
github.com/redis/rueidis.(*call).do(...)
    /go/pkg/mod/github.com/redis/rueidis@v1.0.69/cluster.go:217
github.com/redis/rueidis.(*call).LazyDo.func1(...)
    /go/pkg/mod/github.com/redis/rueidis@v1.0.69/singleflight.go:58
github.com/redis/rueidis.(*call).LazyDo(...)
    /go/pkg/mod/github.com/redis/rueidis@v1.0.69/singleflight.go:53
```

### Client Code

- Connection dial timeout: default
- Connection write timeout: default

#### Client Initialization

```
client, err := rueidisotel.NewClient(
    rueidis.ClientOption{
        InitAddress:       conf.NodeAddresses,
        Username:          conf.UserName,
        Password:          conf.Password,
        DisableCache:      conf.DisableCache,
        CacheSizeEachConn: cacheSizeEachConnection,
    },
)
if err != nil {
    return err
}

if err := client.Do(
    context.Background(),
    client.B().Ping().Build(),
).Error(); err != nil {
    return err
}
```

#### Client Usage

```
func (c *Cache) MGetWithClientSideCache(
    ctx context.Context,
    keys []string,
) (map[string]string, error) {

    results, err := rueidis.MGetCache(
        c.client,
        ctx,
        30*time.Second,
        keys,
    )
    if err != nil {
        return nil, err
    }

    finalResult := make(map[string]string, len(results))
    for key, result := range results {
        value, err := result.ToString()
        if err != nil && !rueidis.IsRedisNil(err) {
            return nil, err
        }
        finalResult[key] = value
    }

    return finalResult, nil
}
```

### Environment

- Rueidis v1.0.69
- Go 1.23.0
- Redis Cluster v7.4
- Auto-Pipelining enabled

### Questions

1. When does rueidis refresh the cluster topology?
If no refresh interval is configured, what situations cause a topology refresh to happen automatically?

2. Is it normal to see EXEC abort errors during a shard failover?
Should applications expect these errors during cluster changes?

3. What situations can lead to Redis protocol parse errors in rueidis?
For example, can they happen due to connection interruptions or cluster changes?

4. After a protocol parse error, should the client recover on its own or be restarted?

--- a/cluster.go
+++ b/cluster.go
@@ -375,21 +375,33 @@ func parseEndpoint(fallback, endpoint string, port int64) string {
 func parseSlots(slots RedisMessage, defaultAddr string) map[string]group {
 	groups := make(map[string]group, len(slots.values()))
 	for _, v := range slots.values() {
-		master := parseEndpoint(defaultAddr, v.values()[2].values()[0].string(), v.values()[2].values()[1].intlen)
+		values := v.values()
+		if len(values) < 3 {
+			continue
+		}
+		masterValues := values[2].values()
+		if len(masterValues) < 2 {
+			continue
+		}
+		master := parseEndpoint(defaultAddr, masterValues[0].string(), masterValues[1].intlen)
 		if master == "" {
 			continue
 		}
 		g, ok := groups[master]
 		if !ok {
 			g.slots = make([][2]int64, 0)
-			g.nodes = make(nodes, 0, len(v.values())-2)
-			for i := 2; i < len(v.values()); i++ {
-				if dst := parseEndpoint(defaultAddr, v.values()[i].values()[0].string(), v.values()[i].values()[1].intlen); dst != "" {
+			g.nodes = make(nodes, 0, len(values)-2)
+			for i := 2; i < len(values); i++ {
+				nodeValues := values[i].values()
+				if len(nodeValues) < 2 {
+					continue
+				}
+				if dst := parseEndpoint(defaultAddr, nodeValues[0].string(), nodeValues[1].intlen); dst != "" {
 					g.nodes = append(g.nodes, NodeInfo{Addr: dst})
 				}
 			}
 		}
-		g.slots = append(g.slots, [2]int64{v.values()[0].intlen, v.values()[1].intlen})
+		g.slots = append(g.slots, [2]int64{values[0].intlen, values[1].intlen})
 		groups[master] = g
 	}
 	return groups
@@ -419,11 +431,21 @@ func parseShards(shards RedisMessage, defaultAddr string, tls bool) map[string]g
 			if dictHealth := dict["health"]; dictHealth.string() != "online" {
 				continue
 			}
-			port := dict["port"].intlen
-			if tls && dict["tls-port"].intlen > 0 {
-				port = dict["tls-port"].intlen
+			dictEndpoint, ok := dict["endpoint"]
+			if !ok {
+				continue
 			}
-			dictEndpoint := dict["endpoint"]
+			portMsg, portOK := dict["port"]
+			port := portMsg.intlen
+			_, tlsPortOK := dict["tls-port"]
+			if !portOK && !tlsPortOK {
+				continue
+			}
+			if tls {
+				if tlsPort, tlsPortOK := dict["tls-port"]; tlsPortOK && tlsPort.intlen > 0 {
+					port = tlsPort.intlen
+				}
+			}
 			if dst := parseEndpoint(defaultAddr, dictEndpoint.string(), port); dst != "" {
 				if dictRole := dict["role"]; dictRole.string() == "master" {
 					m = len(g.nodes)
@@ -431,7 +453,7 @@ func parseShards(shards RedisMessage, defaultAddr string, tls bool) map[string]g
 				g.nodes = append(g.nodes, NodeInfo{Addr: dst})
 			}
 		}
-		if m >= 0 {
+		if m >= 0 && len(g.nodes) > 0 {
 			g.nodes[0], g.nodes[m] = g.nodes[m], g.nodes[0]
 			groups[g.nodes[0].Addr] = g
 		}

--- a/cluster.go
+++ b/cluster.go
@@ -431,21 +431,11 @@ func parseShards(shards RedisMessage, defaultAddr string, tls bool) map[string]g
 			if dictHealth := dict["health"]; dictHealth.string() != "online" {
 				continue
 			}
-			dictEndpoint, ok := dict["endpoint"]
-			if !ok {
-				continue
+			port := dict["port"].intlen
+			if tls && dict["tls-port"].intlen > 0 {
+				port = dict["tls-port"].intlen
 			}
-			portMsg, portOK := dict["port"]
-			port := portMsg.intlen
-			_, tlsPortOK := dict["tls-port"]
-			if !portOK && !tlsPortOK {
-				continue
-			}
-			if tls {
-				if tlsPort, tlsPortOK := dict["tls-port"]; tlsPortOK && tlsPort.intlen > 0 {
-					port = tlsPort.intlen
-				}
-			}
+			dictEndpoint := dict["endpoint"]
 			if dst := parseEndpoint(defaultAddr, dictEndpoint.string(), port); dst != "" {
 				if dictRole := dict["role"]; dictRole.string() == "master" {
 					m = len(g.nodes)
@@ -453,7 +443,7 @@ func parseShards(shards RedisMessage, defaultAddr string, tls bool) map[string]g
 				g.nodes = append(g.nodes, NodeInfo{Addr: dst})
 			}
 		}
-		if m >= 0 && len(g.nodes) > 0 {
+		if m >= 0 {
 			g.nodes[0], g.nodes[m] = g.nodes[m], g.nodes[0]
 			groups[g.nodes[0].Addr] = g
 		}

--- a/cluster_test.go
+++ b/cluster_test.go
@@ -6014,6 +6014,123 @@ func TestClusterClientReplicaOnly_PickMasterIfNoReplica(t *testing.T) {
 	})
 }
 
+func TestClusterSlotsParsing(t *testing.T) {
+	defer ShouldNotLeak(SetupLeakDetection())
+	t.Run("existing fixtures", func(t *testing.T) {
+		for _, test := range []struct {
+			name     string
+			resp     RedisMessage
+			expected map[string]group
+		}{
+			{
+				name: "single shard",
+				resp: slotsResp.val,
+				expected: map[string]group{
+					"127.0.0.1:0": {
+						nodes: nodes{{Addr: "127.0.0.1:0"}, {Addr: "127.0.1.1:1"}},
+						slots: [][2]int64{{0, 16383}},
+					},
+				},
+			},
+			{
+				name: "multi shard",
+				resp: slotsMultiResp.val,
+				expected: map[string]group{
+					"127.0.0.1:0": {
+						nodes: nodes{{Addr: "127.0.0.1:0"}, {Addr: "127.0.1.1:1"}},
+						slots: [][2]int64{{0, 8192}},
+					},
+					"127.0.2.1:0": {
+						nodes: nodes{{Addr: "127.0.2.1:0"}, {Addr: "127.0.3.1:1"}},
+						slots: [][2]int64{{8193, 16383}},
+					},
+				},
+			},
+			{
+				name: "without replicas",
+				resp: slotsMultiRespWithoutReplicas.val,
+				expected: map[string]group{
+					"127.0.0.1:0": {
+						nodes: nodes{{Addr: "127.0.0.1:0"}},
+						slots: [][2]int64{{0, 8192}},
+					},
+					"127.0.1.1:0": {
+						nodes: nodes{{Addr: "127.0.1.1:0"}},
+						slots: [][2]int64{{8193, 16383}},
+					},
+				},
+			},
+		} {
+			t.Run(test.name, func(t *testing.T) {
+				result := parseSlots(test.resp, "127.0.0.1:5")
+				if !reflect.DeepEqual(result, test.expected) {
+					t.Fatalf("unexpected result %#v", result)
+				}
+			})
+		}
+	})
+
+	t.Run("skips malformed entries", func(t *testing.T) {
+		resp := slicemsg('*', []RedisMessage{
+			slicemsg('*', []RedisMessage{}),
+			slicemsg('*', []RedisMessage{
+				{typ: ':', intlen: 0},
+				{typ: ':', intlen: 1},
+				slicemsg('*', []RedisMessage{
+					strmsg('+', "127.0.9.1"),
+				}),
+			}),
+			slicemsg('*', []RedisMessage{
+				{typ: ':', intlen: 2},
+				{typ: ':', intlen: 3},
+				slicemsg('*', []RedisMessage{
+					strmsg('+', "127.0.0.1"),
+					{typ: ':', intlen: 0},
+					strmsg('+', ""),
+				}),
+				slicemsg('*', []RedisMessage{
+					strmsg('+', "127.0.1.1"),
+				}),
+				slicemsg('*', []RedisMessage{
+					strmsg('+', "127.0.1.2"),
+					{typ: ':', intlen: 2},
+					strmsg('+', ""),
+				}),
+			}),
+		})
+		expected := map[string]group{
+			"127.0.0.1:0": {
+				nodes: nodes{{Addr: "127.0.0.1:0"}, {Addr: "127.0.1.2:2"}},
+				slots: [][2]int64{{2, 3}},
+			},
+		}
+
+		result := parseSlots(resp, "127.0.0.1:5")
+		if !reflect.DeepEqual(result, expected) {
+			t.Fatalf("unexpected result %#v", result)
+		}
+	})
+
+	t.Run("malformed reply returns no groups", func(t *testing.T) {
+		resp := slicemsg('*', []RedisMessage{
+			slicemsg('*', []RedisMessage{}),
+			slicemsg('*', []RedisMessage{
+				{typ: ':', intlen: 0},
+			}),
+			slicemsg('*', []RedisMessage{
+				{typ: ':', intlen: 0},
+				{typ: ':', intlen: 1},
+				slicemsg('*', []RedisMessage{}),
+			}),
+		})
+
+		result := parseSlots(resp, "127.0.0.1:5")
+		if len(result) != 0 {
+			t.Fatalf("unexpected result %#v", result)
+		}
+	})
+}
+
 func TestClusterShardsParsing(t *testing.T) {
 	defer ShouldNotLeak(SetupLeakDetection())
 	t.Run("master selection", func(t *testing.T) {
@@ -6071,6 +6188,42 @@ func TestClusterShardsParsing(t *testing.T) {
 			if len(group.nodes) == 0 || group.nodes[0].Addr != master {
 				t.Fatalf("unexpected first node %v", group)
 			}
+		}
+	})
+
+	t.Run("skips nodes missing endpoint or port", func(t *testing.T) {
+		resp := slicemsg(typeArray, []RedisMessage{
+			slicemsg(typeMap, []RedisMessage{
+				strmsg(typeBlobString, "slots"),
+				slicemsg(typeArray, []RedisMessage{
+					strmsg(typeBlobString, "0"),
+					strmsg(typeBlobString, "16383"),
+				}),
+				strmsg(typeBlobString, "nodes"),
+				slicemsg(typeArray, []RedisMessage{
+					slicemsg(typeMap, []RedisMessage{
+						strmsg(typeBlobString, "endpoint"),
+						strmsg(typeBlobString, "127.0.0.1"),
+						strmsg(typeBlobString, "role"),
+						strmsg(typeBlobString, "master"),
+						strmsg(typeBlobString, "health"),
+						strmsg(typeBlobString, "online"),
+					}),
+					slicemsg(typeMap, []RedisMessage{
+						strmsg(typeBlobString, "port"),
+						{typ: typeInteger, intlen: 0},
+						strmsg(typeBlobString, "role"),
+						strmsg(typeBlobString, "master"),
+						strmsg(typeBlobString, "health"),
+						strmsg(typeBlobString, "online"),
+					}),
+				}),
+			}),
+		})
+
+		result := parseShards(resp, "127.0.0.1:5", false)
+		if len(result) != 0 {
+			t.Fatalf("unexpected result %#v", result)
 		}
 	})
 }

--- a/cluster_test.go
+++ b/cluster_test.go
@@ -6190,42 +6190,6 @@ func TestClusterShardsParsing(t *testing.T) {
 			}
 		}
 	})
-
-	t.Run("skips nodes missing endpoint or port", func(t *testing.T) {
-		resp := slicemsg(typeArray, []RedisMessage{
-			slicemsg(typeMap, []RedisMessage{
-				strmsg(typeBlobString, "slots"),
-				slicemsg(typeArray, []RedisMessage{
-					strmsg(typeBlobString, "0"),
-					strmsg(typeBlobString, "16383"),
-				}),
-				strmsg(typeBlobString, "nodes"),
-				slicemsg(typeArray, []RedisMessage{
-					slicemsg(typeMap, []RedisMessage{
-						strmsg(typeBlobString, "endpoint"),
-						strmsg(typeBlobString, "127.0.0.1"),
-						strmsg(typeBlobString, "role"),
-						strmsg(typeBlobString, "master"),
-						strmsg(typeBlobString, "health"),
-						strmsg(typeBlobString, "online"),
-					}),
-					slicemsg(typeMap, []RedisMessage{
-						strmsg(typeBlobString, "port"),
-						{typ: typeInteger, intlen: 0},
-						strmsg(typeBlobString, "role"),
-						strmsg(typeBlobString, "master"),
-						strmsg(typeBlobString, "health"),
-						strmsg(typeBlobString, "online"),
-					}),
-				}),
-			}),
-		})
-
-		result := parseShards(resp, "127.0.0.1:5", false)
-		if len(result) != 0 {
-			t.Fatalf("unexpected result %#v", result)
-		}
-	})
 }
 
 // https://github.com/redis/rueidis/issues/543

